### DISCUSSION TITLE
Feasibility In Providing NO SQL Support for Apollo

### DISCUSSION THREAD

**Discussion by LordMaduz** [apolloconfig/apollo#4759] [Ideas]
Would it be feasible to support NO SQL database like Couchbase as the main Database in Apollo? 
Are there any future plans to provide No SQL support as well?

I believe this would bring major architectural changes to Apollo, So I would like to receive more suggestions and ideas regarding this.

**Comment by nobodyiam** [comment]
There is an [attempt](https://github.com/apolloconfig/apollo/issues/4751) to support databases like h2 and postgres. However, there is currently no discussion of support NO SQL database like Couchbase yet.
Is there any particular reason why a NO SQL database is preferred?

**Comment by hezhangjian** [comment]
Overall, I think running on NoSQL is feasible, but we need to abstract storage and break down different implementations. It's likely to be in a fairly long alpha stage, as we're still heavily reliant on features of relational databases. What is your most pressing demand at the moment? :)

**Comment by LordMaduz** [comment]
The most recent project we are working on relies on heavily unstructured data and after database benchmarking, we have chosen to go with Couchbase for the overall architecture of the system. 
As for the configuration management, we believe apollo would be the best option we have, but as for the banking regulations we are unable to get approval for 2 Databases of SQL and Non-SQL. 
This is the crucial problem we are facing right now.

**Comment by CliffordMalsamlAW** [comment]
Apollo currently supports several SQL databases, including PostgreSQL, MySQL, and SQLite. However, it is not currently designed to work with NoSQL databases like Couchbase out of the box.

In order to support Couchbase or other NoSQL databases, significant changes to the Apollo architecture would be required. This could include modifications to the schema and query systems, as well as the addition of new modules to support NoSQL-specific features like document-oriented data models and dynamic schemas.

While it is technically feasible to support NoSQL databases in Apollo, it would require a significant amount of effort and development time. If this is something you're interested in pursuing, I would recommend starting by researching the differences between SQL and NoSQL databases, and considering how those differences might impact the Apollo architecture.

If you do decide to pursue NoSQL support in Apollo, you may want to consider contributing to the open-source project or working with the Apollo team to develop a solution that meets your needs. Additionally, you could look into building a custom integration with Couchbase or other NoSQL databases, using Apollo as a gateway layer to handle authentication, data validation, and other common features.
